wagtail/admin/views/pages/edit.py

```diff
@@ -79,15 +81,46 @@ def send_commenting_notifications(self):
         if not new_comments and not deleted_comments and not resolved_comments:
             return
 
-        # Get subscribers
+        relevant_comment_ids = [comment.pk for comment in new_comments + resolved_comments]
+
+        # Get global page comment subscribers
         subscribers = PageSubscription.objects.filter(page=self.page, comment_notifications=True).select_related('user')
-        recipient_users = [subscriber.user for subscriber in subscribers if subscriber.user != self.request.user]
+        global_recipient_users = [subscriber.user for subscriber in subscribers if subscriber.user != self.request.user]
+
+        # Get subscribers to individual threads
+        replies = CommentReply.objects.filter(comment_id__in=relevant_comment_ids)
+        comments = Comment.objects.filter(id__in=relevant_comment_ids)
+        thread_users = get_user_model().objects.exclude(pk=self.request.user.pk).exclude(pk__in=subscribers.values_list('user_id', flat=True)).prefetch_related(
+            Prefetch('comment_replies', queryset=replies),
+            Prefetch('comments', queryset=comments)
+        ).exclude(
+            Q(comment_replies__isnull=True) & Q(comments__isnull=True)
+        )
 
         # Skip if no recipients
-        if not recipient_users:
+        if not (global_recipient_users or thread_users):
             return
+        thread_users = [(user, set(list(user.comment_replies.values_list('comment_id', flat=True)) + list(user.comments.values_list('pk', flat=True)))) for user in thread_users]
+        mailed_users = set()
+
+        for current_user, current_threads in thread_users:
```

Comment:
This is really nice! But I think it's worth commenting here that we are trying to avoid calling ``send_notification`` for each user for performance reasons

Reply:
Thanks! Will add a comment